### PR TITLE
BBD-585: bind tag metadata to action creators for use in tracking info

### PR DESCRIPTION
**Pull request by effervescentia**

### CODE REVIEW

File: src/core/action-creator.ts [hunk 1]
```diff
@@ -7,262 +7,266 @@ import Adapters from './adapters';
...
+      // response action creators
+      receiveSearchResponse: (results: Results): Actions.Thunk<any> =>
+        (dispatch, getState) => {
+          const updates = [];
+          const state = getState();
+          if (results.redirect) {
+            updates.push(dispatch(actions.receiveRedirect(results.redirect)));
+          }
+          const recordCount = Adapters.Search.extractRecordCount(results);
+          updates.push(
```
Comment (Vic-Dev) [inline src/core/action-creator.ts]: Should we always send all of these actions? For example, should we send a receiveNavigation action when a user updates the page size (navigations haven't changed)?

File: src/core/action-creator.ts [hunk 2]
```diff
@@ -7,262 +7,266 @@ import Adapters from './adapters';
...
+export default function createActions(flux: FluxCapacitor) {
+
+  return (meta: () => any) => {
+    const metadata = meta();
+    const actions = ({
+      refreshState: (state: any): Actions.RefreshState =>
+        action(Actions.REFRESH_STATE, state, metadata),
+
+      soFetching: (requestType: keyof Store.IsFetching): Actions.SoFetching =>
+        action(Actions.SO_FETCHING, requestType, metadata),
```
Comment (Vic-Dev) [inline src/core/action-creator.ts]: It's probably time to rename this to `IS_FETCHING` :<

File: src/core/actions.ts [hunk 3]
```diff
@@ -1,42 +1,161 @@
...
+  export type ReceivePage = Action<typeof RECEIVE_PAGE, Payload.Page>;
   export const RECEIVE_RECORD_COUNT = 'RECEIVE_RECORD_COUNT';
+  export type ReceiveRecordCount = Action<typeof RECEIVE_RECORD_COUNT, number>;
   export const RECEIVE_TEMPLATE = 'RECEIVE_TEMPLATE';
+  export type ReceiveTemplate = Action<typeof RECEIVE_TEMPLATE, Store.Template>;
   export const RECEIVE_REDIRECT = 'RECEIVE_REDIRECT';
+  export type ReceiveRedirect = Action<typeof RECEIVE_REDIRECT, string>;
 
   // request status
   export const SO_FETCHING = 'SO_FETCHING';
```
Comment (Vic-Dev) [inline src/core/actions.ts]: `IS_FETCHING`

File: src/core/actions.ts [hunk 4]
```diff
@@ -1,42 +1,161 @@
...
+  export type SoFetching = Action<typeof SO_FETCHING, keyof Store.IsFetching>;
```
Comment (Vic-Dev) [inline src/core/actions.ts]: `IS_FETCHING`

File: src/core/reducers/is-fetching.ts [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
+  | Actions.ReceiveAutocompleteProducts
+  | Actions.ReceiveDetailsProduct
+  | Actions.ReceiveProducts;
 export type State = Store.IsFetching;
 
-export default function updateIsFetching(state: State = {}, action): State {
+export default function updateIsFetching(state: State = {}, action: Action): State {
   switch (action.type) {
-    case Actions.SO_FETCHING: return soFetching(state, action);
+    case Actions.SO_FETCHING: return soFetching(state, action.payload);
```
Comment (Vic-Dev) [inline src/core/reducers/is-fetching.ts]: `IS_FETCHING`

File: test/unit/core/reducers/is-fetching.ts [hunk 6]
```diff
@@ -18,7 +17,6 @@ suite('isFetching', ({ expect }) => {
     autocompleteProducts,
     details
   };
-  beforeEach(() => actions = new ActionCreator(<any>{}, <any>{}));
 
   describe('updateIsFetching()', () => {
     it('should update state on SO_FETCHING', () => {
```
Comment (Vic-Dev) [inline test/unit/core/reducers/is-fetching.ts]: `IS_FETCHING`

File: test/unit/core/reducers/is-fetching.ts [hunk 7]
```diff
@@ -31,7 +29,7 @@ suite('isFetching', ({ expect }) => {
         details
       };
 
-      const reducer = isFetching(state, { type: Actions.SO_FETCHING, requestType: 'autocompleteSuggestions' });
+      const reducer = isFetching(state, { type: Actions.SO_FETCHING, payload: 'autocompleteSuggestions' });
```
Comment (Vic-Dev) [inline test/unit/core/reducers/is-fetching.ts]: `IS_FETCHING`

File: src/core/action-creator.ts (hunk 2 above)
Comment (effervescentia) [inline src/core/action-creator.ts]: lmao I was thinking that the other day

File: src/core/action-creator.ts (hunk 1 above)
Comment (effervescentia) [inline src/core/action-creator.ts]: we should change the calls to `thunk()` into calls to `conditional()`, then on a per-action basis we can determine whether or not a change needs to occur. Easy ones area recordCount, page size, some others, navigations is probably the most complicated of all of them. In my mind we can add these piecemeal to improve performance as we go

File: test/unit/core/action-creator.ts [hunk 8]
```diff
@@ -431,44 +438,110 @@ suite('ActionCreator', ({ expect, spy, stub }) => {
...
+          const search: Actions.Payload.Search = { query: 'harry' };
           const dispatch = spy();
 
           actions.updateSearch(search)(dispatch);
 
-          expect(dispatch).to.be.calledWith({ type: Actions.UPDATE_SEARCH, ...<any>search });
+          expect(dispatch).to.be.calledWith({ type: Actions.UPDATE_SEARCH, payload: search, metadata: {} });
         });
 
         it('should not create an UPDATE_SEARCH action when query string is empty', () => {
```
Comment (Vic-Dev) [inline test/unit/core/action-creator.ts]: Or could expect fail on `UPDATE_SEARCH`

File: test/unit/core/action-creator.ts [hunk 9]
```diff
@@ -431,44 +438,110 @@ suite('ActionCreator', ({ expect, spy, stub }) => {
...
           const dispatch = spy();
 
           actions.updateSearch(search)(dispatch);
 
-          expect(dispatch).to.be.calledWith({ type: Actions.UPDATE_SEARCH, query: 'h  a  r\tr  y' });
+          // tslint:disable-next-line max-line-length
+          expect(dispatch).to.be.calledWith({ type: Actions.UPDATE_SEARCH, payload: { query: 'h  a  r\tr  y' }, metadata: {} });
         });
 
         it('should not create an UPDATE_SEARCH action when query string only contains whitespace', () => {
```
Comment (Vic-Dev) [inline test/unit/core/action-creator.ts]: Similar to above

File: test/unit/core/action-creator.ts [hunk 10]
```diff
@@ -431,44 +438,110 @@ suite('ActionCreator', ({ expect, spy, stub }) => {
...
+
+          actions.search();
+
+          expect(updateSearch).to.be.calledWith({ query, clear: true });
+          expect(selectQuery).to.be.calledWith(state);
+        });
+      });
+
+      describe('resetRecall()', () => {
+        it('should call actions.updateSearch() with no params', () => {
```
Comment (Vic-Dev) [inline test/unit/core/action-creator.ts]: That's not what this test says lol

File: test/unit/core/store.ts [hunk 11]
```diff
@@ -27,4 +45,64 @@ suite('Store', ({ expect, stub }) => {
...
+        const storeInstance = { h: 'i' };
+        const initialState = stub(Adapter, 'initialState').returns(state);
+        const applyMiddleware = stub(redux, 'applyMiddleware').returns(middleware);
+        const createStore = stub(redux, 'createStore').returns(storeInstance);
+
+        const store = Store.create(config);
+
+        expect(store).to.eq(storeInstance);
+        expect(initialState).to.be.calledWith(config);
+        expect(applyMiddleware).to.be.calledWithExactly(thunkMiddleware, sinon.match.func, sinon.match.func);
```
Comment (Vic-Dev) [inline test/unit/core/store.ts]: Should we be using calledWithExactly more?

File: test/unit/core/utils.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      const metadata = { c: 'd' };
+      const state: any = { e: 'f' };
+      const thunk = utils.conditional(predicate, ACTION, payload, metadata);
+
+      thunk(() => expect.fail(), () => state);
+
+      expect(predicate).to.be.calledWith(state);
+    });
+
+    it('should return a thunk that dispatches an action if the predicate evaluates to a truthy value', () => {
```
Comment (Vic-Dev) [inline test/unit/core/utils.ts]: Need opposite test?

File: src/core/actions.ts [hunk 13]
```diff
@@ -1,230 +1,174 @@
...
+        state: object;
+      }
     }
 
-    export interface ReceiveSuggestions extends Base {
-      suggestions: string[];
-      categoryValues: string[];
-      navigations: Store.Autocomplete.Navigation[];
+    export namespace Collection {
+      export interface Count {
```
Comment (effervescentia) [inline src/core/actions.ts]: because it's an object that represents the count for a specific collection

File: test/unit/core/action-creator.ts [hunk 14]
```diff
@@ -390,9 +394,11 @@ suite('ActionCreator', ({ expect, spy, stub }) => {
...
           const state = { a: 'b' };
           const allMeta = 'hey';
-          const response = { c: 'd', records: [{ allMeta }] };
+          const record = { allMeta };
+          const response = { c: 'd', records: [record] };
           const detailsProduct = '10293';
           const dispatch = spy();
+          const emit = flux.emit = spy();
           const getStore = spy(() => state);
           const search = stub().resolves(response);
```
Comment (effervescentia) [inline test/unit/core/action-creator.ts]: how do I inline response here?

File: test/unit/core/store.ts (hunk 11 above)
Comment (effervescentia) [inline test/unit/core/store.ts]: Yeah, I think so, not sure why the default isn't "exactly"

File: test/unit/core/utils.ts (hunk 12 above)
Comment (effervescentia) [inline test/unit/core/utils.ts]: covered by the previous test, reworded that test to be more clear

File: test/unit/core/action-creator.ts (hunk 14 above)
Comment (Vic-Dev) [inline test/unit/core/action-creator.ts]: put `const search = stub().resolves({ c: 'd', records: [record] })` bc response isnt used anywhere else. but it doesn't really matter

File: test/unit/core/store.ts (hunk 11 above)
Comment (Vic-Dev) [inline test/unit/core/store.ts]: yeah true

File: test/unit/core/utils.ts (hunk 12 above)
Comment (Vic-Dev) [inline test/unit/core/utils.ts]: nice

**Pull request merged**